File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows = len(board)
        cols = len(board[0]) 

        def backtracking(i, j, word_index):
            if word_index == len(word):
                return True

            if i < 0 or i >= rows or j < 0 or j >= cols or board[i][j] != word[word_index]:
                return False 
            
            temp_mark = board[i][j]
            board[i][j] = "."

            found = (
                backtrack(i + 1, j, word_index + 1) or
                backtrack(i - 1, j, word_index + 1) or
                backtrack(i, j + 1, word_index + 1) or
                backtrack(i, j - 1, word_index + 1)
            )
            
            # Restore the cell
            board[i][j] = temp_mark
            
            return found

        return backtracking(0,0,0)

File: 0079-word-search/0079-word-search.java
class Solution {
    private String word;
    private int rows;
    private int cols;
    private char[][] board;
    
    public boolean exist(char[][] board, String word) {
        this.rows = board.length;   
        this.cols = board[0].length;
        this.word = word;
        this.board = board;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (board[i][j] == word.charAt(0)) {
                    if (backtracking(i, j, 0)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }
    public boolean backtracking(int i, int j, int wordIndex) {
        if (wordIndex == word.length()) return true;
        if (
            i < 0 || 
            i >= rows || 
            j < 0 || 
            j >= cols || 
            board[i][j] != word.charAt(wordIndex)
        ) return false; 
        

        char temp = board[i][j];
        board[i][j] = '.';
    
        boolean found = (
            backtracking(i + 1, j, wordIndex + 1) ||
            backtracking(i - 1, j, wordIndex + 1) ||
            backtracking(i, j + 1, wordIndex + 1) ||
            backtracking(i, j - 1, wordIndex + 1)
        );

        board[i][j] = temp;
        return found;
    }
}
